### PR TITLE
fix: retry module download once if server errored

### LINKED ISSUE #19970: Downloading dependencies returns 500 internal server errors from https://deno.land
Hello. We've received random errors when downloading dependencies when deploying our services on render.com with Docker.

```
ul 28 10:37:43 AM  #11 37.23 error: Import 'https://deno.land/std@0.132.0/node/timers.ts' failed: 500 Internal Server Error
Jul 28 10:37:43 AM  #11 37.23     at https://deno.land/std@0.132.0/node/internal/stream_base_commons.ts:35:33
Jul 28 10:37:43 AM  ------
```

We have to retry each time we see this error and hope it works; it feels very flaky, and this diminishes the dev experience.

Has someone had this error? How can I debug it and fix it?

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes https://github.com/denoland/deno/issues/17251
Closes #19970

### CODE REVIEW

**Comment by piscisaureus** [comment]
Awesome :)
It would be nice if a transient error [here](https://github.com/denoland/deno/pull/17252/files#diff-1b04410b1560deb46a180dc81caff892f21a5ccfc889024552ecda12e167379cL683) (this might be a transport level error such as `ConnectionReset`) would also trigger a retry.

**Comment by bartlomieju** [comment]
> Awesome :) It would be nice if a transient error [here](https://github.com/denoland/deno/blob/782937aa476d9dee1123de95e588308ced2beb3b/cli/file_fetcher.rs#L701) (this might be a transport level error such as `ConnectionReset`) would also trigger a retry.

@piscisaureus this link is not working for me, could you point to relevant line in code?

File: cli/file_fetcher.rs [hunk 1]
```diff
@@ -383,44 +383,74 @@ impl FileFetcher {
...
+        let result = match fetch_once(
+          &client,
+          FetchOnceArgs {
+            url: specifier.clone(),
+            maybe_accept: maybe_accept.clone(),
+            maybe_etag: maybe_etag.clone(),
+            maybe_auth_token: maybe_auth_token.clone(),
+            maybe_progress_guard: maybe_progress_guard.as_ref(),
+          },
+        )
```
Comment (bartlomieju) [inline cli/file_fetcher.rs:400]: Any idea how to structure this code better?

File: cli/file_fetcher.rs [hunk 2]
```diff
@@ -383,44 +383,74 @@ impl FileFetcher {
...
+              Err(generic_error(format!(
+                "Import '{}' failed: {}",
+                specifier, err
+              )))
+            }
+          }
+          FetchOnceResult::ServerError(status) => {
+            // Retry once, and bail otherwise.
+            if !retried {
+              retried = true;
```
Comment (dsherret) [inline cli/file_fetcher.rs:441]: Should we do a slight delay when retrying? (maybe a few ms)

File: test_util/src/lib.rs [hunk 3]
```diff
@@ -808,6 +808,11 @@ async fn main_server(
       *res.status_mut() = StatusCode::FOUND;
       Ok(res)
     }
+    (_, "/server_error") => {
+      let mut res = Response::new(Body::empty());
+      *res.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
+      Ok(res)
+    }
```
Comment (dsherret) [inline test_util/src/lib.rs:815]: Should we add an endoint that errors the first time and succeeds the second?

File: cli/file_fetcher.rs [hunk 4]
```diff
@@ -383,44 +383,74 @@ impl FileFetcher {
...
+              continue;
```
Comment (dsherret) [inline cli/file_fetcher.rs:442]: Perhaps we should `log::warn` when this occurs?

File: test_util/src/lib.rs (hunk 3 above)
Comment (bartlomieju) [reply test_util/src/lib.rs:815]: It's gonna be hard to add state to that server to verify how many times this function was called...

File: cli/file_fetcher.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/file_fetcher.rs:441]: Added 50ms delay.

File: cli/file_fetcher.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/file_fetcher.rs:442]: Done

Comment (bartlomieju) [reply cli/file_fetcher.rs:442]: Now that I'm thinking about it - a flake will cause this to be printed which will still fail tests. Maybe it should be `log::debug` instead?

**Review by dsherret — APPROVED** [review]
LGTM. This will be a big improvement.

**Pull request merged**